### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by Howuhh**

### CODE REVIEW

File: algorithms/small_scale/bc_chaotic_lstm.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    )
+    if config.checkpoints_path is not None:
+        print(f"Checkpoints path: {config.checkpoints_path}")
+        os.makedirs(config.checkpoints_path, exist_ok=True)
+        with open(os.path.join(config.checkpoints_path, "config.yaml"), "w") as f:
+            pyrallis.dump(config, f)
+
+    set_seed(config.train_seed)
+
+    def env_fn():
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py:255]: then we have to add character as an argument, it's not very nice

File: algorithms/small_scale/bc_chaotic_lstm.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+        copy=False
+    )
+    buffer = SequentialBuffer(
+        dataset=tmp_env.get_dataset(mode=config.data_mode, scale="small"),
+        seq_len=config.seq_len,
+        batch_size=config.batch_size,
+        seed=config.train_seed,
+        add_next_step=False
+    )
+    tp = ThreadPoolExecutor(max_workers=14)
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py:276]: fixed, now as config value

File: katakomba/env.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            [tuple] The seeds supplied, in the form (core, disp, reseed).
+        """
+        return self.env.seed(core, disp, reseed)
+
+    def get_normalized_score(self, score: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
+        """
+         Returns score normalized against AutoAscend bot scores achieved for this exact character.
+        """
+        if self.character.count("-") != 2:
+            raise ValueError("Reference score not provided for this character.")
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply katakomba/env.py:146]: fixed

File: algorithms/small_scale/bc_chaotic_lstm.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+def set_seed(seed: int):
+    os.environ["PYTHONHASHSEED"] = str(seed)
+    np.random.seed(seed)
+    random.seed(seed)
+    torch.manual_seed(seed)
+
+
+@torch.no_grad()
+def filter_wd_params(model: nn.Module):
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py:72]: done

File: algorithms/small_scale/bc_chaotic_lstm.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+            continue
+        if 'weight' in name and 'norm' not in name and 'bn' not in name:
+            decay.append(param)
+        else:
+            no_decay.append(param)
+    assert len(no_decay) + len(decay) == len(list(model.parameters()))
+    return no_decay, decay
+
+
+def dict_to_tensor(data, device):
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py:85]: done

File: algorithms/small_scale/bc_chaotic_lstm.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+                    episode_depths.append(infos[i]["current_depth"])
+                    episode_counts[i] += 1
+                    pbar.update(1)
+
+                    current_rewards[i] = 0
+                    current_lengths[i] = 0
+
+    pbar.close()
+    result = {
+        "reward_median": np.median(episode_rewards),
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py:218]: to late, we have all logs in wandb in this format....

[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py:218]: it is consistent across algorithms tho

File: katakomba/utils/datasets/large_scale.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        db_path: str,
+        seq_len: int,
+        batch_size: int,
+        num_workers: int = 8,
+        role: Optional[Role] = None,
+        race: Optional[Race] = None,
+        align: Optional[Alignment] = None,
+        **kwargs
+) -> nld.TtyrecDataset:
+    if not nld.db.exists(db_path):
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply katakomba/utils/datasets/large_scale.py:20]: agree

File: katakomba/utils/datasets/small_scale.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+import numpy as np
+
+from katakomba.utils.roles import Role, Race, Alignment
+from tqdm.auto import tqdm
+
+DATA_PATH = os.environ.get('KATAKOMBA_DATA_DIR', os.path.expanduser('~/.katakomba/datasets'))
+CACHE_PATH = os.environ.get('KATAKOMBA_CACHE_DIR', os.path.expanduser('~/.katakomba/cache'))
+
+
+def _flush_to_memmap(filename: str, array: np.ndarray):
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply katakomba/utils/datasets/small_scale.py:34]: fixed

File: katakomba/utils/datasets/small_scale.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        mmap = np.load(filename, mmap_mode="r")
+    else:
+        mmap = np.memmap(filename, mode="w+", dtype=array.dtype, shape=array.shape)
+        mmap[:] = array
+        mmap.flush()
+
+    return mmap
+
+
+def load_nld_aa_small_dataset(
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply katakomba/utils/datasets/small_scale.py:45]: fixed

File: katakomba/utils/datasets/small_scale.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        return self.data[gameid]
+
+    def __len__(self):
+        return len(self.gameids)
+
+    def metadata(self, idx):
+        gameid = self.gameids[idx]
+        return dict(self.hdf5_file[gameid].attrs)
+
+    def close(self):
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply katakomba/utils/datasets/small_scale.py:134]: fixed

File: katakomba/utils/datasets/small_scale_buffer.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
+import random
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply katakomba/utils/datasets/small_scale_buffer.py:1]: fixed

File: katakomba/utils/datasets/small_scale.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        self.hdf5_file.close()
+        # remove memmap files from the disk upon closing
+        if self.mode == "memmap":
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply katakomba/utils/datasets/small_scale.py]: fixed

File: algorithms/small_scale/bc_chaotic_lstm.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+            "screen_image": torch.tensor(obs["screen_image"][:, None], device=device),
+            "prev_actions": torch.tensor(obs["prev_actions"][:, None], dtype=torch.long, device=device)
+        }
+        logits, new_state = self(inputs, state)
+        actions = torch.argmax(logits.squeeze(1), dim=-1)
+        return actions.cpu().numpy(), new_state
+
+
+@torch.no_grad()
+def vec_evaluate(vec_env, actor, num_episodes,  seed=0, device="cpu"):
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py]: fixed

File: algorithms/small_scale/bc_chaotic_lstm.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+    assert len(no_decay) + len(decay) == len(list(model.parameters()))
+    return no_decay, decay
+
+
+def dict_to_tensor(data, device):
+    return {k: torch.as_tensor(v, device=device) for k, v in data.items()}
+
+
+class Actor(nn.Module):
+    def __init__(self, action_dim, rnn_hidden_dim=512, rnn_layers=1, rnn_dropout=0.0, use_prev_action=True):
```
[the remark this answers is not in the document]
Comment (Howuhh) [reply algorithms/small_scale/bc_chaotic_lstm.py:91]: done

**Review by vkurenkov — APPROVED** [review]

**Pull request merged**